File: 1072-flip-columns-for-maximum-number-of-equal-rows/1072-flip-columns-for-maximum-number-of-equal-rows.py
class Solution:
    def maxEqualRowsAfterFlips(self, matrix: List[List[int]]) -> int:
        hashMap = dict()
        
        for i in matrix:
            
            row_pattern = "".join("T" if num == i[0] else "F" for num in i)
            
            hashMap[row_pattern] = (hashMap.get(row_pattern, 0) + 1)
        
        return max(hashMap.values(), default = 0)
            

File: 1072-flip-columns-for-maximum-number-of-equal-rows/1072-flip-columns-for-maximum-number-of-equal-rows.java
class Solution {
    public int maxEqualRowsAfterFlips(int[][] matrix) {
        Map<String, Integer> map = new HashMap<>();
        
        for(int[] curr : matrix){
            StringBuilder patternBuilder = new StringBuilder("");
            
            for(int col = 0; col < curr.length; col++){
                if(curr[0] == curr[col]){
                    patternBuilder.append("T");
                }else{
                    patternBuilder.append("F");
                }
            }
            
            String rowPattern = patternBuilder.toString();
            map.put(rowPattern, map.getOrDefault(rowPattern, 0) + 1);
        }
        
        int maxFreq = 0;
        
        for(int freq : map.values()){
            maxFreq = Math.max(freq, maxFreq);
        }
        
        return maxFreq;
    }
}